src/compiler/checker.ts

```diff
@@ -5860,6 +5888,70 @@ namespace ts {
             return links.resolvedType;
         }
 
+        /**
+         * Since the source of spread types are object literals, which are not binary,
+         * this function should be called in a left folding style, with left = previous result of getSpreadType
+         * and right = the new element to be spread.
+         */
+        function getSpreadType(left: Type, right: Type, symbol: Symbol): ResolvedType | IntrinsicType {
+            Debug.assert(!!(left.flags & (TypeFlags.Object | TypeFlags.Any)) && !!(right.flags & (TypeFlags.Object | TypeFlags.Any)), "Only object types may be spread.");

```

Comment:
I did, and it would be simple to restore the code I had there earlier. The reason I decided against is that it gets harder to issue a correct, understandable error in `checkObjectLiteral` because intersections that contain type parameters are still not allowed. And I assume that we'll have a good story for all kinds of type operators shortly after 2.1 ships, so a restriction on intersections won't have lasting problems.